rust/operator/src/hdfs_controller.rs

```diff
@@ -0,0 +1,678 @@
+        )
+        .add_data(
+            LOG4J_PROPERTIES.to_string(),
+            hdfs.spec.log4j.as_ref().unwrap_or(&"".to_string()),
+        )
+        .build()
+        .map_err(|source| Error::BuildRoleGroupConfig {
+            source,
+            role: rolegroup_ref.role.clone(),
+            role_group: rolegroup_ref.role_group.clone(),
+        })
+}
+
+fn rolegroup_statefulset(
+    hdfs: &HdfsCluster,
+    rolegroup_ref: &RoleGroupRef<HdfsCluster>,
+    namenode_podrefs: &[HdfsPodRef],
+    journalnode_podrefs: &[HdfsPodRef],
+    hadoop_container: &Container,
+) -> HdfsOperatorResult<StatefulSet> {
+    tracing::info!("Setting up StatefulSet for {:?}", rolegroup_ref);
+    let replicas;
+    let command;
+    let service_name = rolegroup_ref.object_name();
+
+    let mut init_containers = Some(vec![Container {
+        name: "chown-data".to_string(),
+        image: Some(hdfs.hdfs_image()?),
+        args: Some(vec![
+            "sh".to_string(),
+            "-c".to_string(),
+            " chown -R stackable:stackable /data && chmod -R a=,u=rwX /data".to_string(),
+        ]),
+        security_context: Some(SecurityContext {
+            run_as_user: Some(0),
+            ..SecurityContext::default()
+        }),
+        ..hadoop_container.clone()
+    }]);
+
+    let mut containers = vec![];
+
+    let role = serde_yaml::from_str(&rolegroup_ref.role).unwrap();
+    match role {
+        HdfsRole::DataNode => {
+            replicas = hdfs.rolegroup_datanode_replicas(rolegroup_ref)?;
+            command = vec![
+                "/stackable/hadoop/bin/hdfs".to_string(),
+                "--debug".to_string(),
+                "datanode".to_string(),
+            ];
+            init_containers
+                .get_or_insert_with(Vec::new)
+                .extend([Container {
+                    name: "wait-for-namenodes".to_string(),
+                    image: Some(
+                        "docker.stackable.tech/stackable/tools:0.1.0-stackable0".to_string(),
+                    ),
+                    args: Some(vec![
+                        "sh".to_string(),
+                        "-c".to_string(),
+                        format!(
+                            "for jn in {}; do until curl --fail --connect-timeout 2 $jn; do sleep 2; done; done",
+                            namenode_podrefs
+                                .iter()
+                                .map(|pr| format!(
+                                    "http://{}:{}",
+                                    pr.fqdn(),
+                                    pr.ports.get(SERVICE_PORT_NAME_METRICS).map_or(DEFAULT_NAME_NODE_METRICS_PORT, |p| *p)
+                                ))
+                                .collect::<Vec<String>>()
+                                .join(" ")
+                        ),
+                    ]),
+                    ..hadoop_container.clone()
+                }]);
+        }
```

Comment:
I would implement stuff like command, init_containers, containers, ports etc on a per role basis.

Reply:
Yeah, this has become a mess ...